src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs

```diff
@@ -0,0 +1,265 @@
+        public ConcurrencyLimiter(ConcurrencyLimiterOptions options)
+        {
+            _options = options;
+            _permitCount = _options.PermitLimit;
+        }
+
+        /// <inheritdoc/>
+        public override int GetAvailablePermits() => _permitCount;
+
+        /// <inheritdoc/>
+        protected override RateLimitLease AcquireCore(int permitCount)
+        {
+            // These amounts of resources can never be acquired
+            if (permitCount > _options.PermitLimit)
+            {
+                throw new ArgumentOutOfRangeException(nameof(permitCount), $"{permitCount} permits exceeds the permit limit of {_options.PermitLimit}.");
+            }
+
+            // Return SuccessfulLease or FailedLease to indicate limiter state
+            if (permitCount == 0)
+            {
+                return _permitCount > 0 ? SuccessfulLease : FailedLease;
+            }
+
+            // Perf: Check SemaphoreSlim implementation instead of locking
+            if (_permitCount >= permitCount)
+            {
+                lock (Lock)
+                {
+                    if (TryLeaseUnsynchronized(permitCount, out RateLimitLease? lease))
+                    {
+                        return lease;
+                    }
+                }
+            }
+
+            return FailedLease;
+        }
+
+        /// <inheritdoc/>
+        protected override ValueTask<RateLimitLease> WaitAsyncCore(int permitCount, CancellationToken cancellationToken = default)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // These amounts of resources can never be acquired
+            if (permitCount > _options.PermitLimit)
+            {
+                throw new ArgumentOutOfRangeException(nameof(permitCount), $"{permitCount} permits exceeds the permit limit of {_options.PermitLimit}.");
+            }
+
+            // Return SuccessfulLease if requestedCount is 0 and resources are available
+            if (permitCount == 0 && _permitCount > 0)
+            {
+                return new ValueTask<RateLimitLease>(SuccessfulLease);
+            }
+
+            // Perf: Check SemaphoreSlim implementation instead of locking
+            lock (Lock)
+            {
+                if (TryLeaseUnsynchronized(permitCount, out RateLimitLease? lease))
+                {
+                    return new ValueTask<RateLimitLease>(lease);
+                }
+
+                // Don't queue if queue limit reached
+                if (_queueCount + permitCount > _options.QueueLimit)
+                {
+                    // Perf: static failed/successful value tasks?
+                    return new ValueTask<RateLimitLease>(QueueLimitLease);
+                }
+
+                TaskCompletionSource<RateLimitLease> tcs = new TaskCompletionSource<RateLimitLease>(TaskCreationOptions.RunContinuationsAsynchronously);
+                CancellationTokenRegistration ctr = default;
```

Comment:
Added `= default;` since the compiler doesn't like you using non-initialized variables.